scripts/hooks/pre-commit

```diff
@@ -0,0 +1,135 @@
+#!/bin/sh
+#-------------------------------------------------------------------------------
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#-------------------------------------------------------------------------------
+#
+# Called by "git commit" with no arguments.  The hook should
+# exit with non-zero status after issuing an appropriate message if
+# it wants to stop the commit.
+
+RETURN="0"
+SOURCE_FILES_PATTERNS=(
+  # Exclude documentation.
+  ':!*.md'
+  ':!**/README'
+
+  # Exclude data/input files.
+  ':!*.arcs'
+  ':!*.facts'
+  ':!*.csv'
+  ':!*.txt'
+
+  # Exclude test files.
+  ':!*.authlogic'
+  ':!*.raksha'
+  ':!*.ir'
+
+  # Exclude (generated) build and config files.
+  ':(exclude,glob).*'
+  ':!*.bazel'
+  ':!*.yaml'
+  ':(exclude,top)WORKSPACE'
+  ':(exclude,top)Dockerfile'
+
+  # Exclude special case files and directories.
+  ':(exclude,top)build_defs'
+  ':(exclude,top)third_party'
+  ':(exclude,top)images'
+  ':(exclude,top)gcb'
+  ':(exclude,top)rust'
+)
+
+error() {
+  RETURN="1"
+  echo -e "$1"
+}
+
+check_file_for_missing_license() {
+  if [ -d $file ]; then
+    return # This is a directory
+  fi
+  header=$(grep -E "https?:.*LICENSE" $file)
+  if [ ! "$header" == "" ]; then
+    # TODO: Check that the year of the license is correct.
+    return # There was a LICENSE found.
+  fi
+  generated_file=$(grep -E "^// GENERATED FILE, DO NOT EDIT!$" $file)
+  if [ ! "$generated_file" == "" ]; then
+    return # This is a generated file and should not be updated.
+  fi
+  year=$(git log "$file" | head -n 1 | cut -d' ' -f 6)
+  if [ -z $year ]; then
+    year="<unknown>"
+  fi
+  error "MISSING LICENSE in $file  Last change recorded in $year"
```

Comment:
Ah! Thanks, I'll find the first check in then. Thanks

Reply:
> Interesting; why check for this rather than the following?
> 
> ```
> Copyright 20[0-9[0-9] Google LLC
> ```


I thought that would be better at first, but we have code licensed to other authors (e.g. Bazel).

I think you're right and I should switch to checking each of a set of accepted licenses rather than trying to match them all at once.

Reply:
Ah, the files that have different licenses are exclusively in special cased files already. I've switched to using the regex you suggested!